pkg/cloud/services/network/natgateways.go

```diff
@@ -341,3 +341,39 @@ func (s *Service) getNatGatewayForSubnet(sn *infrav1.SubnetSpec) (string, error)
 
 	return "", errors.Errorf("no nat gateways available in %q for private subnet %q, current state: %+v", sn.AvailabilityZone, sn.GetResourceID(), azGateways)
 }
+
+func (s *Service) findNatGatewayForEdgeSubnet(sn *infrav1.SubnetSpec) (string, error) {
+	if sn.IsPublic {
+		return "", errors.Errorf("cannot find NAT gateway for a public subnet, got id %q", sn.GetResourceID())
+	}
+
+	// Check if public edge subnet in the edge zone has nat gateway
+	azGateways := make(map[string][]string)
+	for _, psn := range s.scope.Subnets().FilterPublic() {
+		if psn.NatGatewayID == nil {
+			continue
+		}
+		azGateways[psn.AvailabilityZone] = append(azGateways[psn.AvailabilityZone], *psn.NatGatewayID)
+	}
+
+	if gws, ok := azGateways[sn.AvailabilityZone]; ok && len(gws) > 0 {
+		return gws[0], nil
+	}
+
+	// Check if the parent zone public subnet has nat gateway
+	if sn.ParentZoneName != nil {
+		if gws, ok := azGateways[aws.StringValue(sn.ParentZoneName)]; ok && len(gws) > 0 {
+			return gws[0], nil
+		}
+	}
```

Comment:
@mtulio i have a couple of questions on this block:
- should we error out if we don't find a gateway associated with a ParentZoneName?
- I dont see where azGateways gets initialized with a key for the ParentZoneName.  Am I missing that?

Reply:
Hey @rvanderp3 :
> should we error out if we don't find a gateway associated with a ParentZoneName?
This is part of the design/fallback strategy. If there is not GW for the ParentZone, it will fallback to the first gateway found (line 374)

>  I dont see where azGateways gets initialized with a key for the ParentZoneName. Am I missing that?
`azGAteways` are populated in the line [356][356] with NatGateway IDs for all public subnets.

[]: https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/4882/files#diff-23c065949155c8a08cd81009d31c08a2d86909abe59c908a374d11c9a641281eR356